tests/misc-utils.c

```diff
@@ -0,0 +1,295 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright 2016 Kinvolk GmbH
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ *
+ * Author: Krzesimir Nowak <krzesimir@kinvolk.io>
+ */
+
+#include "misc-utils.h"
+
+#include <errno.h>
+
+#include <unistd.h>
+
+gboolean
+rm_rf (GFile *topdir,
+       GError **error)
+{
+  GQueue queue = G_QUEUE_INIT;
+  GList *dir_stack = NULL;
+  g_autoptr(GError) local_error = NULL;
+  g_autoptr(GFileInfo) top_info;
+
+  top_info = g_file_query_info (topdir,
+                                G_FILE_ATTRIBUTE_STANDARD_TYPE,
+                                G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS,
+                                NULL,
+                                &local_error);
+  if (top_info == NULL)
+    {
+      if (g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+        return TRUE;
+      g_propagate_error (error, g_steal_pointer (&local_error));
+      return FALSE;
+    }
+  if (g_file_info_get_file_type (top_info) != G_FILE_TYPE_DIRECTORY)
+    return g_file_delete (topdir, NULL, error);
+
+  g_queue_push_head (&queue, g_object_ref (topdir));
+  dir_stack = g_list_prepend (dir_stack, g_object_ref (topdir));
+  while (!g_queue_is_empty (&queue))
```

Comment:
Is it worth perhaps making this a recursive function and going depth-first instead of doing it breadth-first like it is now? The upside is that you don't need to keep separate list of directories to delete later. I can imagine that you might possibly want to avoid infinite recursion where there are links involved but I suppose those can be ignored or just deleted directly.